### PR TITLE
ROB: Improve tolerance to invalid Names/Dests

### LINKED ISSUE #1652: "IndexError: list index out of range" while merging a specific PDF document
Merging a specific non encrypted PDF document throws an "IndexError: list index out of range" - this works flawless for other documents but this one document I can not merge. A similar error occurs while reading this document (without error) and than writing again. Providing ```strict=True``` does not result into any additional output.

## Environment

Which environment were you using when you encountered the problem?

```cmd
C:\>python -V
Python 3.8.10

C:\>python -m platform
Windows-10-10.0.19045-SP0

C:\>python -c "import pypdf;print(pypdf.__version__)"
3.4.1
```

## Code + PDF

This is a minimal, complete example that shows the issue:

```python
from pypdf import PdfMerger

merger = PdfMerger()
merger.append(path_to_file)
```

I can't share the PDF publicly as it contains personal data (invoices). Let me know where to send it via email.

## Traceback
```
Traceback (most recent call last):
  File "D:\Workspace\egu\src\egu\blah.py", line 4, in <module>
    merger.append(
  File "D:\Python\virtualenv\egu\lib\site-packages\pypdf\_utils.py", line 441, in wrapper
    return func(*args, **kwargs)
  File "D:\Python\virtualenv\egu\lib\site-packages\pypdf\_merger.py", line 317, in append
    self.merge(len(self.pages), fileobj, outline_item, pages, import_outline)
  File "D:\Python\virtualenv\egu\lib\site-packages\pypdf\_utils.py", line 441, in wrapper
    return func(*args, **kwargs)
  File "D:\Python\virtualenv\egu\lib\site-packages\pypdf\_merger.py", line 217, in merge
    outline = reader.outline
  File "D:\Python\virtualenv\egu\lib\site-packages\pypdf\_reader.py", line 799, in outline
    return self._get_outline()
  File "D:\Python\virtualenv\egu\lib\site-packages\pypdf\_reader.py", line 828, in _get_outline
    self._namedDests = self._get_named_destinations()
  File "D:\Python\virtualenv\egu\lib\site-packages\pypdf\_reader.py", line 756, in _get_named_destinations
    self._get_named_destinations(kid.get_object(), retval)
  File "D:\Python\virtualenv\egu\lib\site-packages\pypdf\_reader.py", line 762, in _get_named_destinations
    value = names[i + 1].get_object()
IndexError: list index out of range
```

### PR DESCRIPTION
**Pull request by pubpub-zz**
fixes #1652

### CODE REVIEW

**Comment by pubpub-zz** [comment]
for testing :
[invalidNamesDest.pdf](https://github.com/py-pdf/pypdf/files/10818844/tt.pdf)

**Comment by pubpub-zz** [comment]
@MartinThoma 
all yours 🙂

File: pypdf/_reader.py [hunk 1]
```diff
@@ -757,9 +757,17 @@ def _get_named_destinations(
         # TABLE 3.33 Entries in a name tree node dictionary (PDF 1.7 specs)
         elif CA.NAMES in tree:  # KIDS and NAMES are exclusives (PDF 1.7 specs p 162)
             names = cast(DictionaryObject, tree[CA.NAMES])
-            for i in range(0, len(names), 2):
+            i = 0
+            while i < len(names):
                 key = cast(str, names[i].get_object())
-                value = names[i + 1].get_object()
```
Comment (MartinThoma) [inline pypdf/_reader.py:762]: Wouldn't it be simpler to  change the line `value = names[i + 1].get_object()` to

```
if len(names) < i + 1:
    break
value = names[i + 1].get_object()
```

Would that work?

Comment (pubpub-zz) [reply pypdf/_reader.py:762]: I wan to deal with cases where I can have [k1 v1 k2 v2 v3 k4 v4 k5 v5] so we need to check/load  items (key or value) one at a time and incrementing i at each step.
I agree that  we can change the code for L766-L769 to:
```
if  i>=len(names):
    break
values = names[i].get_object()
```
I choose my current writing [EAFP](https://docs.python.org/3.6/glossary.html#term-eafp) philosophy

Your recommendation ?

Comment (MartinThoma) [reply pypdf/_reader.py:762]: Thanks for giving the example - I haven't thought about keys / values missing in between. Good job :+1:

**Pull request merged**